### ISSUE TITLE
Add support for tracking XGBoost training metrics by round

### ISSUE THREAD

**Issue by mallman** [allegroai/clearml#381]
Currently ClearML does not track the metrics XGBoost provides via a callback in every training round. It would be very helpful to have these metrics captured automatically.

The callback interface for XGBoost 1.4.2 can be found at https://github.com/dmlc/xgboost/blob/v1.4.2/python-package/xgboost/callback.py.

The Neptune and Polyaxon clients support metric logging from this callback. I can provide specific links into their codebase with details if it would be helpful.

**Comment by bmartinn** [comment]
Hi @mallman 
Sure that sounds like a great idea.
What exactly did you have in mind? Is it automagic, or do you need to provide a callback ?
What can be logged scalars (i.e. over time) or final plots ?
(I have to admit I'm not familiar with the XGBoost auto reporting options)

**Comment by mallman** [comment]
Hi @bmartinn.

I believe it should be possible for this to integration to be "automagical" with some monkey patching. At worst, the user can provide a callback explicitly vended by the task. This is how neptune and polyaxon do it.

XGBoost provides scalar metrics on every round through its callback interface. We should be able to log those and upload when the task is closed. Or, even more magical, perhaps they can be posted to a running task as they are received from xgboost.

My facility with Python is a bit too rudimentary for this to be an easy contribution, though the xgboost callback interface is relatively simple. One complication is that the callback interface has changed significantly with different versions of xgboost. It would be nice to be able to introspect on the currently loaded version of xgboost and monkey patch the right callback.

**Comment by bmartinn** [comment]
> I believe it should be possible for this to integration to be "automagical" with some monkey patching.

Could you point me to an example so that we could test/verify it works (and later add into the examples folder) ?

>  One complication is that the callback interface has changed significantly with different versions of xgboost. It would be nice to be able to introspect on the currently loaded version of xgboost and monkey patch the right callback.

I think this should be doable, but I would start with supporting the latest version, anything I'm missing here ?

**Comment by mallman** [comment]
> Could you point me to an example so that we could test/verify it works (and later add into the examples folder) ?

Here's an example from XGBoost 1.3.0:

https://github.com/dmlc/xgboost/blob/v1.3.0/demo/guide-python/callbacks.py.

I believe this is still the supported API in the latest version.

Also, these unit tests may be helpful:

https://github.com/dmlc/xgboost/blob/v1.4.0/tests/python/test_callback.py

The pre-1.3.0 callback interface appears to be largely undocumented. Perhaps it's best to limit support for callback-based logging to XGBoost 1.3+ at this point and consider support for XGBoost 1.2 for a future update only if there is a real demand.

**Comment by bmartinn** [comment]
Thanks @mallman !
Basically if I understand this example correctly, we want the `Plotting()` callback to happen automagically, correct ?
Just putting it out there , the demo specifically says it is [slow](https://github.com/dmlc/xgboost/blob/1bf389998383f333490155dba4608bff9ca63b42/demo/guide-python/callbacks.py#L18), is this dues to plotly actually rendering the plot, or is it the callback itself? Basically I'm trying to figure out is it okay we will be sending all the scalers every iteration ?

> The pre-1.3.0 callback interface appears to be largely undocumented. Perhaps it's best to limit support for callback-based logging to XGBoost 1.3+ at this point and consider support for XGBoost 1.2 for a future update only if there is a real demand.

Agreed.

**Comment by mallman** [comment]
> Thanks @mallman !
> Basically if I understand this example correctly, we want the `Plotting()` callback to happen automagically, correct ?
> Just putting it out there , the demo specifically says it is [slow](https://github.com/dmlc/xgboost/blob/1bf389998383f333490155dba4608bff9ca63b42/demo/guide-python/callbacks.py#L18), is this dues to plotly actually rendering the plot, or is it the callback itself? Basically I'm trying to figure out is it okay we will be sending all the scalers every iteration ?

I suspect the work the callback does in the example is what's considered "slow"—that is, updating the plots. I cannot imagine calling a function—even if it's using some kind of reflective interface, which I doubt it is—would have a material impact on training performance compared to the actual computational work.

I believe if ClearML uploaded the scalars in a single batch after training that would provide a lot of value for post-hoc analysis and cross-comparison. Realtime or online updates would be useful but not necessary. This is what I'm hearing from our data science team.

I'm thinking if this data is a few floating point numbers per iteration, the upload size should be negligible.

**Comment by mallman** [comment]
Hi @bmartinn.

One of my data science coworkers has implemented an example integration with a custom XGBoost callback. Though this is tied to tensorboard, you might find this useful in considering how to implement an automagical callback. This comes from a couple of JupyterLab cells. He was able to successfully log the training scalars to ClearML this way.

```python
import os
import tempfile

import xgboost as xgb
from tensorboardX import SummaryWriter
from xgboost.callback import TrainingCallback

class TensorBoardCallback(TrainingCallback):
    '''Print the evaluation result at each iteration.
    .. versionadded:: 1.3.0
    Parameters
    ----------
    rank : int
        Which worker should be used for printing the result.
    period : int
        How many epoches between printing.
    '''
    
    def __init__(self, period=1, show_stdv=False) -> None:
        
        self.period = period
        assert period > 0
        # last error message, useful when early stopping and period are used together.
        self._latest: Optional[str] = None
        super().__init__()


    def after_iteration(self, model, epoch, evals_log) -> bool:
        if not evals_log:
            return False

        writer = SummaryWriter(logdir=os.path.join(tempfile.gettempdir(), 'histogram_example'))

        for data, metric in evals_log.items():
            for metric_name, log in metric.items():
                score = log[-1]
                
                if (epoch % self.period) == 0 or self.period == 1:
                    writer.add_scalar(data + '-' + metric_name, score, epoch)

        return False

    def after_training(self, model):
        return model
```

```python
from sklearn.datasets import load_boston
from sklearn.model_selection import train_test_split

X, y = load_boston(return_X_y=True)
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=100)

dtrain = xgb.DMatrix(X_train, label=y_train)
dtest = xgb.DMatrix(X_test, label=y_test)

params = {'objective':'reg:squarederror', 'eval_metric': 'rmse'}

bst = xgb.train(params, dtrain, num_boost_round=100, evals=[(dtrain, 'train'), (dtest, 'test')],
        callbacks=[TensorBoardCallback(period=3)], verbose_eval=0)
```

**Comment by jkhenning** [comment]
Hi @mallman, thanks! We'll take a look at this soon 🙂

**Comment by bmartinn** [comment]
Hi @mallman 

Good news 🎉  we just pushed the XGBoost callback integration (it was stuck in the pipeline for too long ... )

You can test the new example [here](https://github.com/allegroai/clearml/blob/master/examples/frameworks/xgboost/xgboost_metrics.py)

If you can verify it works with your code, that would be great.
Install directly from github with:
```
pip install git+https://github.com/allegroai/clearml.git
```

**Comment by mallman** [comment]
Hi @bmartinn. That's fantastic! Thanks so much for working on this.

We are currently in a transition from a PoC deployment (now retired) to a production deployment. A lot has changed in the Kubernetes deployment process from 1.0.2 (which we ran for the PoC) and 1.1 (which we're deploying for production), so it's going to take some extra time. However, I will certainly report back as soon as I've had a chance to try this in production.

**Comment by bmartinn** [comment]
@Maimonator 
Can you verify that this [example](https://github.com/allegroai/clearml/blob/master/examples/frameworks/xgboost/xgboost_metrics.py) as is, did not work in Jupyter ?
What's the Python / XGboost version you are using?

**Comment by jkhenning** [comment]
Closing this issue. Please reopen if required.
